src/main/org/codehaus/groovy/reflection/ClassInfo.java

```diff
@@ -186,30 +185,20 @@ public void setStrongMetaClass(MetaClass answer) {
         MetaClass strongRef = strongMetaClass;
         
         if (strongRef instanceof ExpandoMetaClass) {
-          ((ExpandoMetaClass)strongRef).inRegistry = false;
-          synchronized(modifiedExpandos){
-            for (Iterator<ClassInfo> it = modifiedExpandos.iterator(); it.hasNext(); ) {
-              ClassInfo info = it.next();
-              if(info == this){
-                it.remove();
-              }
+            ((ExpandoMetaClass)strongRef).inRegistry = false;
+            for (Iterator<ClassInfo> itr = modifiedExpandos.iterator(); itr.hasNext(); ) {
+                ClassInfo info = itr.next();
+                if(info == this) {
+                    itr.remove();
+                }
             }
-          }
         }
 
         strongMetaClass = answer;
 
         if (answer instanceof ExpandoMetaClass) {
-          ((ExpandoMetaClass)answer).inRegistry = true;
-          synchronized(modifiedExpandos){
-            for (Iterator<ClassInfo> it = modifiedExpandos.iterator(); it.hasNext(); ) {
-              ClassInfo info = it.next();
-                if(info == this){
-                  it.remove();
-                }
-             }
-             modifiedExpandos.add(this);
-          }
```

Comment:
Removed because it appeared to be a duplicate of the logic above.  Only if the previous previous `strongMetaClass` value was an expando would it have been added to the map, so the iteration/removal above should have taken care to remove this `ClassInfo` from the `modifiedExpandos` and there's no reason to iterate again.